generative/networks/nets/cdpm.py

```diff
@@ -0,0 +1,589 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+
+import math
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from generative.networks.layers.RPE import RPEAttention
+from monai.networks.blocks import Convolution
+
+
+class TimestepBlock(nn.Module):
+    """
+    Any module where forward() takes timestep embeddings as a second argument.
+    """
+
+    @abstractmethod
+    def forward(self, x, emb):
+        """
+        Apply the module to `x` given `emb` timestep embeddings.
+        """
+
+
+class TimestepEmbedAttnThingsSequential(nn.Sequential, TimestepBlock):
+    """
+    A sequential module that passes extra things to the children that
+    support it as an extra input.
+    """
+    def forward(self, x, emb, attn_mask, T=1, frame_indices=None)-> torch.Tensor:
+        for layer in self:
+            if isinstance(layer, TimestepBlock):
+                kwargs = dict(emb=emb)
+                kwargs['emb'] = emb
+            elif isinstance(layer, FactorizedAttentionBlock):
+                kwargs = dict(
+                    temb=emb,
+                    attn_mask=attn_mask,
+                    T=T,
+                    frame_indices=frame_indices,
+                )
+            else:
+                kwargs = {}
+            x = layer(x, **kwargs)
+        return x
+
+
+class Upsample(nn.Module):
+    """
+    An upsampling layer with an optional convolution.
+    :param channels: channels in the inputs and outputs.

```

Comment:
Please change Sphinx style tags to google style tags, here and elsewhere in the code.